### PR TITLE
Extend model repository API for config override

### PR DESCRIPTION
**Pull request by GuanLuo**
@deadeyegoodwin @tanmayv25 @CoderHam Look and feel of the extended API. In terms of GRPC API, I think we could `ModelRepositoryParameter` extend to have `ModelConfig` as one of the value option, but requiring JSON string for now to match the HTTP side.

### CODE REVIEW

File: docs/protocol/extension_model_repository.md [hunk 1]
```diff
@@ -116,8 +116,29 @@ $repository_index_error_response =
...
+- "parameters" : An object containing zero or more parameters for this
+  request expressed as key/value pairs. See
+  [Parameters](https://github.com/kserve/kserve/blob/master/docs/predict-api/v2/required_api.md#parameters)
+  for more information.
+
+The load API accepts the following parameters:
+
+- "config" : string parameter that contains a JSON representation of the model
+configuration, which must be able to be parsed into [ModelConfig message from
+model_config.proto](https://github.com/triton-inference-server/common/blob/main/protobuf/model_config.proto). This config will be used for loading the model instead of the one in
```
Comment (deadeyegoodwin) [inline docs/protocol/extension_model_repository.md:140]: format long lines to match existing

**Review by deadeyegoodwin — CHANGES_REQUESTED** [review]

**Review by deadeyegoodwin — APPROVED** [review]

File: docs/protocol/extension_model_repository.md [hunk 2]
```diff
@@ -116,8 +116,30 @@ $repository_index_error_response =
...
+```
+
+- "parameters" : An object containing zero or more parameters for this
+  request expressed as key/value pairs. See
+  [Parameters](https://github.com/kserve/kserve/blob/master/docs/predict-api/v2/required_api.md#parameters)
+  for more information.
+
+The load API accepts the following parameters:
+
+- "config" : string parameter that contains a JSON representation of the model
```
Comment (CoderHam) [inline docs/protocol/extension_model_repository.md:138]: IF a load operation is called on a model with a specified config, what will be the expected behavior?

File: docs/protocol/extension_model_repository.md [hunk 3]
```diff
@@ -303,3 +336,10 @@ message RepositoryModelUnloadResponse
 {
 }
 ```
+
+The RepositoryModelUnload API accepts the following parameters:
+
+- "unload_dependents" : boolean parameter indicating that in addition
+  to unloading the requested model, also unload any dependent model
+  that was loaded along with the requested model (for example, the
```
Comment (CoderHam) [inline docs/protocol/extension_model_repository.md:344]: For example, the ensemble models that were composed of the requested model.

**Review by CoderHam — CHANGES_REQUESTED** [review]

File: docs/protocol/extension_model_repository.md (hunk 2 above)
Comment (GuanLuo) [reply docs/protocol/extension_model_repository.md:138]: "This config will be used for loading the model instead of the one in
the model directory."

Comment (tanmayv25) [reply docs/protocol/extension_model_repository.md:138]: @GuanLuo If the load request with a config parameter is issued over a model that is already loaded, my guess is the model with previous config would be unloaded and the re-loaded model will use the config from the request. Am I right? 

May be just nitpicking. If this happens on an ensemble model where the config parameter modifies the composing model. The load API can trigger multiple model unloads as well. May be it is a good idea to call out these cases explicitly in the documentation.

File: docs/protocol/extension_model_repository.md [hunk 4]
```diff
@@ -116,8 +116,32 @@ $repository_index_error_response =
 
 The load API requests that a model be loaded into Triton, or reloaded
 if the model is already loaded. A load request is made with an HTTP
-POST to a load endpoint. The HTTP body must be empty. A successful
-load request is indicated by a 200 HTTP status.
+POST to a load endpoint.  The HTTP body may be empty or may contain
```
Comment (CoderHam) [inline docs/protocol/extension_model_repository.md:119]: Extra space before `The HTTP`

File: src/core/model_repository_manager.cc [hunk 5]
```diff
@@ -1221,8 +1221,13 @@ ModelRepositoryManager::Create(
     // model loading / unloading error will be printed but ignored
     RETURN_IF_ERROR(local_manager->PollAndUpdateInternal(&all_models_polled));
   } else {
+    std::unordered_map<std::string, std::vector<const InferenceParameter*>>
+        models;
+    for (const auto& model_name : startup_models) {
+      models[model_name];
```
Comment (CoderHam) [inline src/core/model_repository_manager.cc:1227]: Is this typo?

File: src/servers/http_server.cc [hunk 6]
```diff
@@ -1265,7 +1265,48 @@ HTTPAPIServer::HandleRepositoryControl(
...
+            HTTP_RESPOND_IF_ERR(req, param.AsString(&config));
+            auto param = TRITONSERVER_ParameterNew(
+                "config", TRITONSERVER_PARAMETER_STRING, config.c_str());
+            if (param != nullptr) {
+              params.emplace_back(param);
+            } else {
+              HTTP_RESPOND_IF_ERR(
+                  req, TRITONSERVER_ErrorNew(
+                           TRITONSERVER_ERROR_INTERNAL,
+                           "unexpected error on creating Triton parameter"));
```
Comment (CoderHam) [inline src/servers/http_server.cc:1303]: Instead of "unexpected error on creating Triton parameter", maybe we can report a more verbose error?

**Review by CoderHam — CHANGES_REQUESTED** [review]

File: src/core/model_repository_manager.cc (hunk 5 above)
Comment (GuanLuo) [reply src/core/model_repository_manager.cc:1227]: This is to create a default map element, which is model to load without any additional load parameters

**Review by CoderHam — APPROVED** [review]

**Pull request merged**